core/runtime/impl/memory_impl.cpp

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "runtime/impl/memory_impl.hpp"
+
+namespace kagome::runtime {
+
+  MemoryImpl::MemoryImpl() {
+    offset_ = 0;
+  }
+
+  MemoryImpl::MemoryImpl(SizeType size) : MemoryImpl() {
+    memory_.resize(size);
+  }
+
+  void MemoryImpl::resize(uint32_t newSize) {
+    // Ensure the smallest allocation is large enough that most allocators
+    // will provide page-aligned storage. This hopefully allows the
+    // interpreter's MemoryImpl to be as aligned as the MemoryImpl being
+    // simulated, ensuring that the performance doesn't needlessly degrade.
+    //
+    // The code is optimistic this will work until WG21's p0035r0 happens.
+    const size_t minSize = 1 << 12;
+    size_t oldSize = memory_.size();
+    memory_.resize(std::max(minSize, static_cast<size_t>(newSize)));
+    if (newSize < oldSize && newSize < minSize) {
+      std::memset(&memory_[newSize], 0, minSize - newSize);
+    }
+  }
+
+  std::optional<Memory::AddressType> MemoryImpl::allocate(SizeType size) {
+    if (size == 0) {
+      return 0;
+    }
+    if (size > memory_.size()) {
```

Comment:
Shouldn't we try to resize our memory in that case in order to allocate in any case?

Reply:
You are right! Wasm linear memory [specification](https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md#linear-memories) actually supports your opinion. It says that memory can grow dynamically

Reply:
Although it can grow up to max memory size which is in case of 32 bit space is 4 Gib